File: carDemo/src_html/app/simulation/car.js


define(
	[
		'../utils/collision.js'
	],
	function(
		utils
	)
{

	//

	var createCar = function (elem_id)
	{
		this._position = {x:0, y:0};
		this._angle = 0;
		this._alive = true;

		this._fitness = 0;

		this._total_update = 0;

		this._trail = [];

		this._updateSensors();
	};

	//

	createCar.prototype.update = function(step, walls, ann)
	{
		if (!this._alive)
			return;

		//

		var input = [];

		for (var i = 0; i < this._sensors.length; ++i)
			input.push( this._sensors[i].value );


		// console.log("input,", input);

		var output = ann.process( input );

		var leftTheta = output[0];
		var rightTheta = output[1];

		// console.log("leftTheta=" + leftTheta);
		// console.log("rightTheta=" + rightTheta);

		leftTheta = Math.min(Math.PI/32.0, Math.max(-Math.PI/32.0, leftTheta));
		rightTheta = Math.min(Math.PI/32.0, Math.max(-Math.PI/32.0, rightTheta));

		this._angle += (leftTheta - rightTheta) * step;

		// console.log(this._angle);

		var speed = ((Math.abs(leftTheta + rightTheta)) / 2) * 160.0;

		speed = Math.min(15.0, Math.max(10.0, speed));

		//

		// var speed = 10.0;

		this._position.x += (speed * Math.cos(this._angle)) * step;
		this._position.y += (speed * Math.sin(this._angle)) * step;

		this._trail.push( {x:this._position.x, y:this._position.y} );

		//

		this._collideWalls( walls );
		this._updateSensors();
		this._collideSensors( walls );
		this._collideCheckpoints();

		++this._total_update;
	};

	//

	createCar.prototype._collideWalls = function(walls)
	{
		for (var i = 0; i < walls.length; ++i)
			if (utils.CollisionSegmentCercle(walls[i].p1, walls[i].p2, this._position, 10.0))
			{
				this._alive = false;
				break;
			}
	}

	//

	createCar.prototype._updateSensors = function(walls)
	{
		// initialise

		var angles = [
			-Math.PI/2.0,
			-Math.PI/4.0,
			0.0,
			Math.PI/4.0,
			Math.PI/2.0
		];

		this._sensors = [];
		for (var i = 0; i < angles.length; ++i)
		{
			this._sensors.push({
				line: {
					  p1: {x:0, y:0}
					, p2: {
						  x: 100.0 * Math.cos(angles[i])
						, y: 100.0 * Math.sin(angles[i])
					}
				}
				, value: 1.0
			});
		}

		// translate/rotate

		for (var i = 0; i < this._sensors.length; ++i)
		{
			var sensor = this._sensors[i];

			sensor.line.p1.x += this._position.x;
			sensor.line.p2.x += this._position.x;
			sensor.line.p1.y += this._position.y;
			sensor.line.p2.y += this._position.y;

			sensor.line.p1 = utils.rotateVec2(sensor.line.p1, this._position, this._angle);
			sensor.line.p2 = utils.rotateVec2(sensor.line.p2, this._position, this._angle);
		}
	}

	//

	createCar.prototype._collideSensors = function(walls)
	{
		for (var i = 0; i < this._sensors.length; ++i)
		{
			var sensor = this._sensors[i];

			for (var j = 0; j < walls.length; ++j)
			{
				var wall = walls[j];

				var v = utils.CollisionSegSeg_partial(sensor.line.p1, sensor.line.p2, wall.p1, wall.p2);

				if (v >= 0.0 && v < 1.0)
					sensor.value = v;
			}
		}
	}

	//

	createCar.prototype._collideCheckpoints = function()
	{
		var i = 0;
		while (i < this._checkpoints.length)
		{
			if (utils.CollisionSegmentCercle(this._checkpoints[i].p1, this._checkpoints[i].p2, this._position, 5.0))
			{
				// console.log('checkpoint erased');
				this._checkpoints.splice(i, 1);
				++this._fitness;
			}
			else
			{
				++i;
			}
		}

		if (this._checkpoints.length == 0)
		{
			// console.log('last checkpoint erased');

			// // this line reward a faster car once the circuit is completed
			this._fitness += (1000.0 / this._total_update);

			this._alive = false;
		}
	}

	//

	return createCar;
})

File: carDemo/src_cpp/src/Experiment/Simulation/logic/Car.cpp


#include "Car.hpp"

#include "Circuit.hpp"

#include "../utils/utils.hpp"
#include "../utils/collision.hpp"


#include <cmath>


Car::Car()
	:	m_angle(0.0f),
		m_fitness(0.0f),
		m_alive(true),
		m_min_updates(50),
		m_total_updates(0),
		m_current_checkpoint(0)
{
	updateSensors();
}

void	Car::update(float step, const Circuit& circuit, const NeuralNetwork& in_NN)
{
	if (std::isnan(m_position.x) ||
		std::isnan(m_position.y) ||
		std::isnan(m_angle))
		m_alive = false;

	if (!m_alive)
		return;

	//
	// min update

	if (m_min_updates > 0)
		--m_min_updates;
	if (m_min_updates == 0)
	{
		// m_min_updates = 50;
		m_alive = false;
	}

	// min update
	//
	//

	// m_fitness += 0.01f;

	this->updateSensors();
	this->collideSensors( circuit.getWalls() );
	this->collideCheckpoints( circuit.getCheckpoints() );
	this->collideWalls( circuit.getWalls() );

	// return;

	//


	std::vector<float>	input;
	std::vector<float>	output;

	for (t_sensor& sensor : m_sensors)
		input.push_back( sensor.m_value );

	// if (input.size() != 5)
	//     throw std::invalid_argument( "about to use an invalid number of inputs" );

	in_NN.process(input, output);

	// if (output.size() != 2)
	//     throw std::invalid_argument( "received invalid number of outputs" );

	float leftTheta		= output[0];
	float rightTheta	= output[1];

	if (std::isnan(leftTheta))	leftTheta = 0.0f;
	if (std::isnan(rightTheta))	rightTheta = 0.0f;



	float steer_range = M_PI/32.0f;
	leftTheta = std::min(steer_range, std::max(-steer_range, leftTheta));
	m_angle += leftTheta * step;

	float speed = rightTheta * 3.0f;
	speed = std::min(15.0f, std::max(-10.0f, speed));





	t_vec2f	prev_pos = m_position;

	m_position.x += (speed * cosf(m_angle)) * step;
	m_position.y += (speed * sinf(m_angle)) * step;


	m_trail.push_back( t_line(prev_pos, m_position) );

	++m_total_updates;
}

void	Car::updateSensors()
{
	// initialise

	// static const float angles[5] = {
	// 	-M_PI/2.0f,
	// 	-M_PI/4.0f,
	// 	0.0f,
	// 	M_PI/4.0f,
	// 	M_PI/2.0f
	// };

	static const float angles[5] = {
		-M_PI/4.0f,
		-M_PI/8.0f,
		0.0f,
		M_PI/8.0f,
		M_PI/4.0f
	};

	for (unsigned int i = 0; i < m_sensors.size(); ++i)
	{
		m_sensors[i].m_line.p1 = t_vec2f();
		m_sensors[i].m_line.p2.x = 200.0f * cosf(angles[i]);
		m_sensors[i].m_line.p2.y = 200.0f * sinf(angles[i]);

		m_sensors[i].m_value = 1.0f;
	}

	// translate/rotate

	for (t_sensor& sensor : m_sensors)
	{
		sensor.m_line.p1.x += m_position.x;
		sensor.m_line.p2.x += m_position.x;
		sensor.m_line.p1.y += m_position.y;
		sensor.m_line.p2.y += m_position.y;

		sensor.m_line.p1 = rotateVec2(sensor.m_line.p1, m_position, m_angle);
		sensor.m_line.p2 = rotateVec2(sensor.m_line.p2, m_position, m_angle);
	}
}

void	Car::collideSensors(const t_lines& walls)
{
	for (t_sensor& sensor : m_sensors)
	{
		sensor.m_value = 1.0f;

		for (const t_line& wall : walls)
		{
			float v = CollisionSegmentSegment_partial(sensor.m_line.p1, sensor.m_line.p2, wall.p1, wall.p2);

			if (v >= 0.0f && v < 1.0f && sensor.m_value > v)
				sensor.m_value = v;
		}
	}
}

void	Car::collideCheckpoints(const t_lines& checkpoints)
{
	if (m_current_checkpoint < checkpoints.size())
	{
		auto& checkpoint = checkpoints[m_current_checkpoint];
		if (CollisionSegmentCercle(checkpoint.p1, checkpoint.p2, m_position, 5.0f))
		{
			m_min_updates = 50;
			++m_fitness;
			++m_current_checkpoint;
		}
	}

	if (m_current_checkpoint >= checkpoints.size())
	{
		// this line reward a faster car once the circuit is completed
		m_fitness += (1000.0f / m_total_updates);

		m_alive = false;
	}
}

void	Car::collideWalls(const t_lines& walls)
{
	for (const t_line& l : walls)
		if (CollisionSegmentCercle(l.p1, l.p2, m_position, 10.0f))
		{
			m_alive = false;
			break;
		}
}

void	Car::reset(const Circuit& circuit)
{
	m_position = circuit.getStartingPositon();
	m_angle = circuit.getStartingAngle();

	m_alive = true;
	m_fitness = 0;
	m_total_updates = 0;
	m_trail.clear();
	m_min_updates = 50;
	m_current_checkpoint = 0;
}

float	Car::getLife() const
{
	return (m_min_updates / 50.0f);
}
